### ISSUE TITLE
Is MDX supported in SvelteKit?

### ISSUE THREAD

**Issue by tsriram** [sdorra/content-collections#251]
Hello there! First of all, thanks much for this great library. I've been using it in a Next.js project with MDX and it's been great. 

I just started with a SvelteKit project and I couldn't get it the MDX content rendered correctly. Based on my understanding from the docs, `MDXContent` component is available only for React. I tried using `@html` directive from Svelte to render MDX content but it's not working. I also see that the [SvelteKit sample](https://github.com/sdorra/content-collections/tree/22b7006bf791b350fb299c1b9d8622b3e92630ee/samples/svelte-kit) only has Markdown example. Can someone please let me know if MDX is fully supported in SvelteKit? Or should I use this along with a library like [MDsvEX](https://github.com/pngwn/MDsveX)? Thanks!

**Comment by sdorra** [comment]
Hi @tsriram, MDX works only with react. MDsveX is the way to go for svelte. It should be possible to use [compile of MDsveX](https://mdsvex.pngwn.io/docs#compile) within the transform function of content-collections.

And if you want to go an extra mile, i would happily accept a pr with `@content-collections/mdsvex` 😁

**Comment by tsriram** [comment]
Got it, thanks @sdorra. Sure, I'd love to contribute. Will get to this in the coming week.

Just to make sure my understanding is correct, `@content-collections/mdsvex` would be the Svelte equivalent of `@content-collections/mdx` right?

**Comment by sdorra** [comment]
Yes, the package should call the compile function of mdsvex and cache the results. You can look at how the MDX and Markdown packages are implemented for inspiration.

**Comment by theetrain** [comment]
I recently started using `content-collections` with mdsvex. This is what a `/src/routes/blog/[slug]/+page.js` looks like on my site:

```js
import { allPosts } from 'content-collections'
import { error } from '@sveltejs/kit'

/** @type {import('./$types').PageLoad} */
export async function load({ params }) {
  const post = allPosts.find((el) => el.slug === params.slug)

  // 🔽 import is handled by mdsvex preprocessor
  const Post = (await import(`$blog/${post._meta.path}.svx`)).default

  if (!post) error(404, 'Post not found.')

  return {
    Post,
    title: post.title,
    slug: post.slug
  }
}
```

Since this leverages the mdsvex Vite plugin, it allows imports in `.svx` files such as:

`/src/blog/hello-world.svx`
```md
---
title: Hello world
---

<script>
  import Thing from '$blog/Thing.svelte'
</script>

<Thing />

Some test markdown.
```

I think things become limiting when using the `compile` function supplied from `mdsvex`, which can translate HTML elements into components, but cannot resolve imports.

Perhaps the ideal solution is to create an adder for `content-https://svelte-add.com/, and add the option to include `mdsvex` boilerplate. On a related note, we can add boilerplate for Shiki integration via the Svelte adder, too.

**Comment by git-no** [comment]
@sdorra There a are some Svelte markdown-preprocessor repositories. Why not integrating an existing markdown-preprocessor code? Or do you want to develop a new markdown-preprocessor for content-collections?

**Comment by sdorra** [comment]
@git-no as far as I understand, `mdsvex` is a popular Markdown preprocessor in the Svelte environment, and the idea is to create an `mdsvex` integration for content collections.

**Comment by git-no** [comment]
@sdorra MDsveX seem to me to oversized in this combination. Also the MDsveX `compiler` does not work quite well.

I tried today to develop a svelteMDx compiler to integrate in transform similar to `compileMDX` but difficulty is, frontmatter (as object or string) is not available and in `document` is only the body of the markdown. So I am handicapped to proceed.

**Comment by sdorra** [comment]
@git-no, the first parameter of the `transform` function is the `document`, and the document contains all fields from the schema, plus the content and the `_meta` field. 

The schema describes the shape of the frontmatter, e.g.:

If we have the following md/mdx file:

```md
---
title: "Functioning of the Drive of the Heart of Gold"
author: "Tricia Marie McMillan"
---

# The Heart of Gold
...
```

The configuration of content collections (`content-collections.ts`) could look like this:

```ts
import { defineCollection, defineConfig } from "@content-collections/core";
import { compileMDX } from "@content-collection/svelte-mdx";

const posts = defineCollection({
  name: "posts",
  directory: "posts",
  include: "*.md(x)?",
  schema: (z) => ({
    title: z.string(),
    author: z.string()
  }),
  transform: async (document, ctx) => {
    const content = await compileMDX(ctx, document);
    return {
      ...document,
      content,
    };
  },
});

export default defineConfig({
  collections: [posts],
});
```

The document that is passed to the `transform` function should look like the following:

 ```json
 {
   "title": "Functioning of the Drive of the Heart of Gold"
   "author": "Tricia Marie McMillan",
   "content": "# The Heart of Gold",
   "_meta": { ... }
 }
 ```
 
So if you remove the `content` and the `_meta`, you have the front matter as an object.

**Comment by git-no** [comment]
I got Svelte Component running in markdown without using MDsveX and overall it is quite simple, without extra dynamic import a markdown file via node or Vite.

That can be achieved with support of the node package `svelte-exmarkdown`.  

Just add Svelte Component into md files, e.g. 

```md
---
title: With Svelte
summary: Markdown and Svelte Components
---

## Counter 
Here is the counter

<Counter></Counter>
```

The `+page.svelte` file

```svelte
<script lang="ts">
  import type { PageProps } from './$types';
  import Markdown from 'svelte-exmarkdown';
  import { markdownPlugins } from '../markdownPlugins';

  let { data }: PageProps = $props();

  const post = $derived(data.post);
  const plugins = [markdownPlugins];
</script>

<h1>{post.title}</h1>
<Markdown md={post.content} {plugins} />
```

> `post` is a record from content-collections generated file.

`+page.ts`
```typescript
import type { PageLoad } from './$types';
import { error } from '@sveltejs/kit';
import { allPosts } from 'content-collections';
import { markdownPlugins } from '../markdownPlugins';

export const load: PageLoad = async ({ params }) => {
  const post = allPosts.find((post) => post.slug == params.slug);
  if (!post) {
    error(404, `Could not find path ${params.slug}`);
  }

  return {
    post,
  };
};
```

The `markdownPlugins.ts` file

```typescript
import type { Plugin } from 'svelte-exmarkdown';
import rehypeRaw from 'rehype-raw';
import Counter from '$lib/Counter.svelte';

export const markdownPlugins: Plugin = {
  rehypePlugin: rehypeRaw,
  renderer: {
    counter: Counter
  }
};
```
Of course more Svelte Components can be added to the renderer object.

In my example the component `Counter.svelte`looks like (Svelte 5)

```svelte
<script>
  let count = $state(0);
</script>

<button onclick={() => count++}>
  clicks: {count}
</button>
```

A second example [Stackblitz](https://stackblitz.com/edit/svelte-exmarkdown-custom-elemtents?file=src%2FApp.svelte&terminal=dev)

@sdorra @tsriram  FYI

**Comment by git-no** [comment]
addendum:

Depending on the writing of the Svelte component within the markdown, in HTML the Svelte component code will inside of an `<p>` tag or not:

#### Inside `<p>`tag

Selfclosing Svelte Component -> rendered inside a <p> tag

```md
<Counter />
```

#### Without `<p>`tag

Closing tag Svelte Component not in same line  -> rendered without `<p>` tag

```md
<Counter>
</Counter>
```

#### Issue

Fails to render markdown after Svelte Component

Closing tag Svelte Component in the same line -> markdown after Svelte component is not in HTML

```md
<Counter></Counter>

lorem ipsum
```
